### PR TITLE
Add sr_retrieve_changes

### PR DESCRIPTION
**Pull request by syyyr**
Fixes: https://github.com/sysrepo/sysrepo/issues/2774

### CODE REVIEW

File: src/sysrepo.c [hunk 1]
```diff
@@ -3462,6 +3462,18 @@ sr_has_changes(sr_session_ctx_t *session)
     return 0;
 }
 
+API int
+sr_retrieve_changes(sr_session_ctx_t *session, const struct lyd_node **changes)
```
Comment (michalvasko) [inline src/sysrepo.c:3466]: Why not just `sr_get_changes()`?

File: src/sysrepo.c [hunk 2]
```diff
@@ -3462,6 +3462,18 @@ sr_has_changes(sr_session_ctx_t *session)
...
+{
+    sr_error_info_t *err_info = NULL;
+
+    SR_CHECK_ARG_APIRET(!session, session, err_info);
+
+    *changes = session->dt[session->ds].edit->tree;
```
Comment (michalvasko) [inline src/sysrepo.c:3472]: `session->dt[session->ds].edit` can be NULL.

File: src/sysrepo.c [hunk 3]
```diff
@@ -3462,6 +3462,18 @@ sr_has_changes(sr_session_ctx_t *session)
     return 0;
 }
 
+API int
```
Comment (michalvasko) [inline src/sysrepo.c:3465]: Would probably just return `const struct lyd_node *`, the error is not interesting.

File: src/sysrepo.h [hunk 4]
```diff
@@ -962,6 +962,15 @@ int sr_nacm_apply_changes(sr_session_ctx_t *session, uint32_t timeout_ms);
  */
 int sr_has_changes(sr_session_ctx_t *session);
 
+/**
+ * @brief Retrieve stored changes (prepared, not yet applied).
```
Comment (michalvasko) [inline src/sysrepo.h:966]: Include a note somewhere that the pointer stops being valid as soon as the changes are applied or discarded.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/sysrepo.c [hunk 5]
```diff
@@ -3462,6 +3462,16 @@ sr_has_changes(sr_session_ctx_t *session)
     return 0;
 }
 
+API const struct lyd_node *
+sr_get_changes(sr_session_ctx_t *session)
+{
+    if (!session->dt[session->ds].edit) {
```
Comment (michalvasko) [inline src/sysrepo.c:3468]: Add `session` NULL check.

File: src/sysrepo.h [hunk 6]
```diff
@@ -962,6 +962,17 @@ int sr_nacm_apply_changes(sr_session_ctx_t *session, uint32_t timeout_ms);
...
+ *
+ * Note: as soon as the changes get applied or discarded, the return value becomes invalid.
+ *
+ * @param[in] session Session ([DS](@ref sr_datastore_t)-specific) to retrieve changes from.
+ * @param[out] changes Where to store the changes.
```
Comment (michalvasko) [inline src/sysrepo.h:971]: Obsolete.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/sysrepo.c [hunk 7]
```diff
@@ -3462,6 +3462,21 @@ sr_has_changes(sr_session_ctx_t *session)
     return 0;
 }
 
+API const struct lyd_node *
+sr_get_changes(sr_session_ctx_t *session)
+{
+    if (!session) {
```
Comment (michalvasko) [inline src/sysrepo.c:3468]: This is fine, I guess, I meant just adding `!session` into the `if` below.

Comment (syyyr) [reply src/sysrepo.c]: I thought the extra log message makes sense to differentiate the errors. Changed.

File: src/sysrepo.c [hunk 8]
```diff
@@ -3462,6 +3462,16 @@ sr_has_changes(sr_session_ctx_t *session)
     return 0;
 }
 
+API const struct lyd_node *
+sr_get_changes(sr_session_ctx_t *session)
+{
+    if (!session && !session->dt[session->ds].edit) {
```
Comment (michalvasko) [inline src/sysrepo.c:3468]: s/&&/||

**Review by michalvasko — CHANGES_REQUESTED** [review]

**Review by michalvasko — APPROVED** [review]

**Pull request merged**